sentry_streams/sentry_streams/adapters/arroyo/broadcaster.py

```diff
@@ -0,0 +1,89 @@
+from collections import deque
+from copy import deepcopy
+from time import time
+from typing import Deque, Optional, Sequence, Union, cast
+
+from arroyo.processing.strategies.abstract import MessageRejected, ProcessingStrategy
+from arroyo.types import FilteredPayload, Message, Value
+
+from sentry_streams.adapters.arroyo.routes import Route, RoutedValue
+
+
+class Broadcaster(ProcessingStrategy[Union[FilteredPayload, RoutedValue]]):
+    """
+    Custom processing strategy which duplicates a message once per downstream branch
+    and updates the waypoints of each copy to correspond to one of the branches.
+    Duplicates keep the timestamp from the original message.
+    """
+
+    def __init__(
+        self,
+        route: Route,
+        downstream_branches: Sequence[str],
+        next_step: ProcessingStrategy[Union[FilteredPayload, RoutedValue]],
+    ) -> None:
+        self.__next_step = next_step
+        self.__route = route
+        self.__downstream_branches = downstream_branches
+        # If we get MessageRejected from the next step, we put the pending messages here
+        self.__pending: Deque[Message[RoutedValue]] = deque()
+
+    def __flush_pending(self) -> None:
+        while self.__pending:
+            try:
+                message = self.__pending[0]
+                self.__next_step.submit(message)
+                self.__pending.popleft()
+            except MessageRejected:
+                pass
```

Comment:
This code will end in an infinite loop if the downstream step is saturated without a way to make progress.
Imagine the downstream step is at capacity.
You will start trying to submit the message till the downstream step keep raising MessageRejected.
Though, as you would never call next.poll in this loop, the downstream step will never make progress and keep raising the exception.